keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -0,0 +1,660 @@
+
+- We plan to allow access to the current state of the object to allow validation rules to check the new value against the current value(e.g. for validation ratcheting or immutability checks).
+
+
+#### Field paths and field patterns
+
+A field path is a patch to a single node in the data tree. I.e. it specifies the
+exact indices of the list items and the keys of map entries it traverses.
+
+A field *pattern* is a path to all nodes in the data tree that match the pattern. I.e.
+it may wildcard list item and map keys.
+
+
+#### Expression lifecycle
+
+When CRDs are written to the kube-apiserver, all expressions will be [parsed and
+typechecked](https://github.com/google/cel-go#parse-and-check) and the resulting
+program will be cached for later evaluation (CEL evaluation is thread-safe and
+side-effect free). Any parsing or type checking errors will cause the CRD write
+to fail with a descriptive error.
+
+#### Function library
+
+The function library available to expressions can be augmented using [extension
+functions](https://github.com/google/cel-spec/blob/master/doc/langdef.md#extension-functions).
+
+List of functions to include for the initial release:
+- Equality and Ordering (customized to do Kuberenetes semantic equality that handles "associative lists")
+- Regular Expressions
+- Some Standard Definitions
+
+
+Considerations:
+- The functions will become VERY difficult to change as this feature matures. We
+  should limit ourselves initially to functions that we have a high level of
+  confidence will not need to be changed or rethought.
+
+- Support kubernetes specific concepts, like accessing associative lists by key may be needed, but
+  we need to review more use cases to determine if this is needed.
+  
+- The Kubernetes associated list equality uses map semantic equality which is different from CEL. 
+  We would consider overwriting in CEL or adding a workaround utility function.
+
+
+### User Stories
+
+- Use case: [Tekton pipeline validation](https://github.com/tektoncd/pipeline/blob/main/pkg/apis/pipeline/v1beta1/pipeline_validation.go)
+  - Referential integrity checks
+  - Custom formatted validation error messages
+  - "Either list X or list Y must be non-empty"
+  - "There exists a X in one list/map for each Y in another map/list"
+- (PRs to add additional user stories to this list welcome!)
+
+### Notes/Constraints/Caveats (Optional)
+
+While we believe the expressiveness of CEL is pretty complete for our purposes, it is non-turing
+complete, and lacks support recursive data types (OpenAPIv3 & CRD structural schemas are not
+possible to validate with CEL).
+
+### Risks and Mitigations
+
+#### Accidental misuse
+
+Break the control plane by consuming excessive CPU and/or memory the api-server.
+
+Mitigation: CEL is specifically designed to constrain the running time of expressions and to limit
+the memory utilization. Since CRD creation is a privileged operation, it should be safe to
+integrate.
```

Comment:
I _think_ CEL gives us sufficient controls here, but we still have to determine the limits to set. Describing how we'll determine those would be useful.

Reply:
👍 Added some comments about publishing performance benchmarks to test CPU and memory limits before going to GA.

Reply:
because of the potential abuse vectors, I think we need to understand upper bounds before enabling by default

Reply:
Added to Beta graduation criteria. Memory is going to be tricky, but maybe we can work with the CEL community. I'll look into that.